### DISCUSSION TITLE
Multiple IKE SAs between two peers

### DISCUSSION THREAD

**Discussion by dsolanki-h9** [strongswan/strongswan#2251] [Q&A]
I have a strongswan config :Status of IKE charon daemon (strongSwan 5.8.2, Linux 5.4.0-1045-aws, x86_64):

carriergw-mnoOperator-00cb608a-b1a0-471b-ba27-72015e69cd95:  172.16.0.240,...66.94.4.170  IKEv2, dpddelay=120s
carriergw-mnoOperator-00cb608a-b1a0-471b-ba27-72015e69cd95:   local:  [XXX.com] uses public key authentication
carriergw-mnoOperator-00cb608a-b1a0-471b-ba27-72015e69cd95:    cert:  "CN=XXX"
carriergw-mnoOperator-00cb608a-b1a0-471b-ba27-72015e69cd95:   remote: uses public key authentication
carriergw-mnoOperator-00cb608a-b1a0-471b-ba27-72015e69cd95:   child:  dynamic === 10.192.0.0/14 TUNNEL, dpdaction=restart

conn carriergw-mnoOperator-00cb608a-b1a0-471b-ba27-72015e69cd95
    auto=start
    #authby=psk
    type=tunnel
    left=172.16.0.240,
    leftsourceip="%config"
    leftid=XXX
    rightsubnet=10.192.0.0/14
    leftauth=pubkey
    leftcert=/etc/ipsec.d/carriergw/certs/mnoOperator-00cb608a-b1a0-471b-ba27-72015e69cd95_cert.pem
    right=66.94.4.170
    rightid="%any"
    #rightsubnet=10.51.0./24
    rightauth=pubkey
    #ike=aes128-aes256-sha1-sha256-modp1024-modp2048-modp4096
    ike=aes256-sha256-modp2048-!
    #esp=aes128-aes256-sha1-sha256-modp1024-modp2048-modp4096
    esp=aes256-sha256-modp2048-!
    closeaction=restart
    dpdaction=restart
    dpddelay=120s
    ikelifetime=28800s
    lifetime=3600s
    #ikelifetime=1h
    #lifetime=8h
    reauth=no
    mobike=no
    leftallowany=no
    keyingtries=%forever

Scenario:
Earlier thr is only single IKE SA with one ipsec SA.
Delete IKE SA is called with 
swanctl --teardown --child carriergw-mnoOperator-00cb608a-b1a0-471b-ba27-72015e69cd95 --timeout 10
& after that IPsec conf is added back & done ipsec update so that again Ipsec should come up.

Logs are:
May 15 04:34:42 ip-10-0-4-56 charon: 15[IKE] deleting IKE_SA carriergw-mnoOperator-00cb608a-b1a0-471b-ba27-72015e69cd95[135] between 172.16.0.240[XXX.highway9.com]...66.94.4.170[ CN=XXX_mocn_segwblv.com]
May 15 04:34:42 ip-10-0-4-56 charon: 15[IKE] sending DELETE for IKE_SA carriergw-mnoOperator-00cb608a-b1a0-471b-ba27-72015e69cd95[135]
May 15 04:34:42 ip-10-0-4-56 charon: 15[ENC] generating INFORMATIONAL request 2 [ D ]
May 15 04:34:42 ip-10-0-4-56 charon: 15[NET] sending packet: from 172.16.0.240[4500] to 66.94.4.170[4500] (80 bytes)
May 15 04:34:42 ip-10-0-4-56 charon: 05[IKE] destroying IKE_SA in state CONNECTING without notification
May 15 04:34:42 ip-10-0-4-56 charon: 12[IKE] destroying IKE_SA in state CONNECTING without notification
May 15 04:34:42 ip-10-0-4-56 charon: 09[NET] received packet: from 66.94.4.170[4500] to 172.16.0.240[4500] (80 bytes)
May 15 04:34:42 ip-10-0-4-56 charon: 09[ENC] parsed INFORMATIONAL response 2 [ ]
May 15 04:34:42 ip-10-0-4-56 charon: 09[IKE] IKE_SA deleted
May 15 04:34:42 ip-10-0-4-56 charon: 09[IKE] DNS server 10.168.251.140 still used, decreasing refcount
May 15 04:34:42 ip-10-0-4-56 charon: 09[IKE] DNS server 10.168.243.36 still used, decreasing refcount
May 15 04:34:42 ip-10-0-4-56 charon: 15[LIB]   file coded in unknown format, discarded
May 15 04:34:42 ip-10-0-4-56 charon: 15[LIB] building CRED_CERTIFICATE - X509 failed, tried 4 builders
May 15 04:34:51 ip-10-0-4-56 charon: 11[KNL] creating delete job for CHILD_SA ESP/0xccaded41/172.16.0.240
May 15 04:34:51 ip-10-0-4-56 charon: 14[JOB] CHILD_SA ESP/0xccaded41/172.16.0.240 not found for delete
May 15 04:34:51 ip-10-0-4-56 charon: 05[IKE] giving up after 5 retransmits
May 15 04:34:51 ip-10-0-4-56 charon: 05[IKE] installing new virtual IP 10.178.179.30
May 15 04:34:51 ip-10-0-4-56 charon: 05[IKE] restarting CHILD_SA carriergw-mnoOperator-00cb608a-b1a0-471b-ba27-72015e69cd95
May 15 04:34:51 ip-10-0-4-56 charon: 05[IKE] initiating IKE_SA carriergw-mnoOperator-00cb608a-b1a0-471b-ba27-72015e69cd95[145] to 66.94.4.170  -----------> >>>>>>>>>>>> 1st IKE
May 15 04:34:51 ip-10-0-4-56 charon: 05[ENC] generating IKE_SA_INIT request 0 [ SA KE No N(NATD_S_IP) N(NATD_D_IP) N(FRAG_SUP) N(HASH_ALG) N(REDIR_SUP) ]
May 15 04:34:51 ip-10-0-4-56 charon: 05[NET] sending packet: from 172.16.0.240[500] to 66.94.4.170[500] (464 bytes)
May 15 04:34:51 ip-10-0-4-56 charon: 05[IKE] removing DNS server 10.168.251.140 via resolvconf
May 15 04:34:52 ip-10-0-4-56 charon: 15[NET] received packet: from 66.94.4.170[500] to 172.16.0.240[500] (1497 bytes)
May 15 04:34:52 ip-10-0-4-56 charon: 15[ENC] parsed IKE_SA_INIT response 0 [ SA KE No N(NATD_S_IP) N(NATD_D_IP) CERTREQ ]
May 15 04:34:52 ip-10-0-4-56 charon: 10[ENC] parsed IKE_AUTH response 1 [ IDr CERT AUTH CPRP(ADDR DNS DNS) SA TSi TSr ]
May 15 04:34:52 ip-10-0-4-56 charon: 10[IKE] authentication of XXX with RSA signature successful
May 15 04:34:52 ip-10-0-4-56 charon: 10[IKE] IKE_SA carriergw-mnoOperator-00cb608a-b1a0-471b-ba27-72015e69cd95[145] established between 172.16.0.240[XXX.highway9.com]...66.94.4.170[CN=XXX_mocn_segwblv.com]
May 15 04:34:52 ip-10-0-4-56 charon: 10[IKE] scheduling rekeying in 2976s
May 15 04:34:52 ip-10-0-4-56 charon: 10[IKE] maximum IKE_SA lifetime 3516s
May 15 04:34:52 ip-10-0-4-56 charon: 10[IKE] installing DNS server 10.168.243.36 via resolvconf
May 15 04:34:52 ip-10-0-4-56 charon: 16[LIB]   file coded in unknown format, discarded
May 15 04:34:52 ip-10-0-4-56 charon: 16[LIB] building CRED_CERTIFICATE - X509 failed, tried 4 builders
May 15 04:34:53 ip-10-0-4-56 charon: 10[IKE] installing DNS server 10.168.251.140 via resolvconf
May 15 04:34:53 ip-10-0-4-56 charon: 05[IKE] retransmit 2 of request with message ID 0
May 15 04:34:53 ip-10-0-4-56 charon: 05[NET] sending packet: from 172.16.0.240[500] to 172.16.0.110[500] (464 bytes)
May 15 04:34:54 ip-10-0-4-56 charon: 10[IKE] installing new virtual IP 10.178.179.30
May 15 04:34:54 ip-10-0-4-56 charon: 10[IKE] CHILD_SA carriergw-mnoOperator-00cb608a-b1a0-471b-ba27-72015e69cd95{834} established with SPIs c7baa856_i 004a6ed0_o and TS 10.178.179.30/32 === 10.192.0.0/14
May 15 04:34:54 ip-10-0-4-56 charon: 10[IKE] establishing CHILD_SA carriergw-mnoOperator-00cb608a-b1a0-471b-ba27-72015e69cd95{835}
May 15 04:34:54 ip-10-0-4-56 charon: 10[ENC] generating CREATE_CHILD_SA request 2 [ SA No KE TSi TSr ]
May 15 04:34:54 ip-10-0-4-56 charon: 10[NET] sending packet: from 172.16.0.240[4500] to 66.94.4.170[4500] (480 bytes)
May 15 04:34:54 ip-10-0-4-56 charon: 07[IKE] initiating IKE_SA carriergw-mnoOperator-00cb608a-b1a0-471b-ba27-72015e69cd95[146] to 66.94.4.170 >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>> 2nd IKE
May 15 04:34:54 ip-10-0-4-56 charon: 07[ENC] generating IKE_SA_INIT request 0 [ SA KE No N(NATD_S_IP) N(NATD_D_IP) N(FRAG_SUP) N(HASH_ALG) N(REDIR_SUP) ]
May 15 04:34:54 ip-10-0-4-56 charon: 07[NET] sending packet: from 172.16.0.240[500] to 66.94.4.170[500] (464 bytes)
May 15 04:34:54 ip-10-0-4-56 charon: 12[NET] received packet: from 66.94.4.170[4500] to 172.16.0.240[4500] (80 bytes)
May 15 04:34:54 ip-10-0-4-56 charon: 12[ENC] parsed CREATE_CHILD_SA response 2 [ N(NO_ADD_SAS) ]
May 15 04:34:54 ip-10-0-4-56 charon: 12[IKE] received NO_ADDITIONAL_SAS notify, no CHILD_SA built
May 15 04:34:54 ip-10-0-4-56 charon: 12[IKE] failed to establish CHILD_SA, keeping IKE_SA
May 15 04:34:54 ip-10-0-4-56 charon: 12[IKE] establishing CHILD_SA carriergw-mnoOperator-00cb608a-b1a0-471b-ba27-72015e69cd95{836}
May 15 04:34:54 ip-10-0-4-56 charon: 12[ENC] generating CREATE_CHILD_SA request 3 [ SA No KE TSi TSr ]
May 15 04:34:54 ip-10-0-4-56 charon: 12[NET] sending packet: from 172.16.0.240[4500] to 66.94.4.170[4500] (480 bytes)
May 15 04:34:54 ip-10-0-4-56 charon: 05[NET] received packet: from 66.94.4.170[500] to 172.16.0.240[500] (1497 bytes)
May 15 04:34:54 ip-10-0-4-56 charon: 05[ENC] parsed IKE_SA_INIT response 0 [ SA KE No N(NATD_S_IP) N(NATD_D_IP) CERTREQ ]

Suscpected logs: 
Not sure when this scenario happens where CHILD SA not found for delete.
May 15 04:34:51 ip-10-0-4-56 charon: 11[KNL] creating delete job for CHILD_SA ESP/0xccaded41/172.16.0.240
May 15 04:34:51 ip-10-0-4-56 charon: 14[JOB] CHILD_SA ESP/0xccaded41/172.16.0.240 not found for delete
May 15 04:34:51 ip-10-0-4-56 charon: 05[IKE] giving up after 5 retransmits
May 15 04:34:51 ip-10-0-4-56 charon: 05[IKE] installing new virtual IP 10.178.179.30
May 15 04:34:51 ip-10-0-4-56 charon: 05[IKE] restarting CHILD_SA carriergw-mnoOperator-00cb608a-b1a0-471b-ba27-72015e69cd95
May 15 04:34:51 ip-10-0-4-56 charon: 05[IKE] initiating IKE_SA carriergw-mnoOperator-00cb608a-b1a0-471b-ba27-72015e69cd95[145] to 66.94.4.170

**Comment by tobiasbrunner** [comment]
> Delete IKE SA is called with
> swanctl --teardown --child carriergw-mnoOperator-00cb608a-b1a0-471b-ba27-72015e69cd95 --timeout 10

That command does not delete the IKE_SA. First the command is called `--terminate`, second with the `--child` option, only the matching CHILD_SA is deleted, the IKE_SA is not. Also, if you have `swanctl` available, why use `ipsec` and `ipsec.conf` and not `swanctl.conf`?

> after that IPsec conf is added back

What does that mean?  Back from where? Was it removed before?

> done ipsec update so that again Ipsec should come up.

Because of `auto=start` you mean? Possible, don't know exactly if that only happens if the config changed. But in the version you use every `ipsec up` or initiation triggered by starter due to `auto=start` (or even by the daemon via `closeaction`/`dpdaction`) could potentially create a duplicate as these were not prevented back then (whether it's just a duplicate CHILD_SA or a duplicate IKE_SA depends on config changes and the `charon.reuse_ikesa` option).

Also note that the IKE_SA you marked as first one in the log is created due to `dpdaction=restart`:
```
May 15 04:34:51 ip-10-0-4-56 charon: 05[IKE] giving up after 5 retransmits
May 15 04:34:51 ip-10-0-4-56 charon: 05[IKE] installing new virtual IP 10.178.179.30
May 15 04:34:51 ip-10-0-4-56 charon: 05[IKE] restarting CHILD_SA carriergw-mnoOperator-00cb608a-b1a0-471b-ba27-72015e69cd95
```


> Not sure when this scenario happens where CHILD SA not found for delete.

That's the allocated inbound SPI expiring in the kernel at the same time the daemon stopped waiting for the last retransmit. The `xfrm_acq_expires` option is set to the total retransmission timeout by default.

**Comment by dsolanki-h9** [comment]
Thanks for the clarification.  As mentioned its mostly ipsec auto=start & dpdaction triggered at the same time causing this.

swanctl --terminate --ike carriergw-mnoOperator-00cb608a-b1a0-471b-ba27-72015e69cd95 --timeout 10
 we are using this command.. Given wrong command previously.

 if you have swanctl available, why use ipsec and ipsec.conf and not swanctl.conf?
>>> We just reusing our old implementation. There is a plan to move to swanctl in future release.

What does that mean? Back from where? Was it removed before?
>>> Yes. After terminating ike SA. We delete the sa config file. and perform ipsec update.
& then again add the ipsec config file & do the ipsec update again.  Configuration has auto=start and due to this IKE SA gets created on ipsec update.

option charon.reuse_ikesa is not set explicitly in our setup.

**Reply by tobiasbrunner** [reply]
> swanctl --terminate --ike carriergw-mnoOperator-00cb608a-b1a0-471b-ba27-72015e69cd95 --timeout 10
> we are using this command. Given wrong command previously.

OK, but note that in the version you are using, even though a delete task will get queued, if the IKE_SA is currently sending retransmits for some other exchange, the SA gets reestablished once the retransmission timeout hits (a check for this was added with 5.8.3, 96b61792dfff1721df64b2e9a5076caef8c7eda0).

> Yes. After terminating ike SA. We delete the sa config file. and perform ipsec update.

That sounds kinda weird. Why not just update the config and then call `ipsec update`? Or even `ipsec reload` if you really want to flush out the old configs? Also note that with swanctl, configs that have `start_action=start` will get terminated automatically if they are unloaded/modified.

**Reply by dsolanki-h9** [reply]
Thanks. Will see when we can update our setups with new version so that duplicate IKE SA wont be observed for the scenario mentioned.

 Why not just update the config and then call ipsec update? Or even ipsec reload if you really want to flush out the old configs? 
>>>> With ipsec reload its mentioned in the document that "All currently established connections could be affected by this "., so on the safer side simply using ipsec update, so that we are sure only that specific connection will be impacted and not the other.
Currently we are deleting and rewriting the config, as thr could be config change if UI system gives different config in our setup.
This is not the good way for configuration,  we will be moving to swanctl way of configuring in future releases. Currently we just used swanctl command to cleanup the SA thats it. That too as ipsec down command was not having timeout option.

**Reply by tobiasbrunner** [reply]
Deleting the file and calling `ipsec update` has exactly the same effect as calling `ipsec reload`, that is, it removes all existing configs, which could cause problems when rekeying existing SAs. If you want to not affect existing SAs that use unmodified configs (those that are removed or modified could still be affected), then you have to leave those in the file when you call `ipsec update`.

> That too as ipsec down command was not having timeout option.

`ipsec down` does not, but there is `ipsec stroke down-nb` which does not wait for the command to complete (it's basically like setting `--timeout` to `-1`).
